jelares/portfolio/src/main/java/com/google/sps/servlets/BlogDataServlet.java

```diff
@@ -0,0 +1,159 @@
+      final String blogPostContent = request.getParameter("blog-post-content");
+      final long timestamp = System.currentTimeMillis();
+
+      // Get the URL of the image that the user uploaded to Blobstore.
+      String blogPostImageUrl = getUploadedFileUrl(request, "blog-post-image");
+
+      // Add the input to datastore
+      Entity blogPostEntity = new Entity("blog-post");
+      blogPostEntity.setProperty("blog-post-title", blogPostTitle);
+      blogPostEntity.setProperty("blog-post-content", blogPostContent);
+      blogPostEntity.setProperty("blog-post-image", blogPostImageUrl);
+      blogPostEntity.setProperty("timestamp", timestamp);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(blogPostEntity);
+
+      // Redirect back to the HTML page.
+      response.sendRedirect("/index.html#blog-box");
+    }
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
+  }
+
+  /** Turns blog posts to json */
+  private String blogToJson(List<BlogPost> blogPosts) {
+    Gson gson = new Gson();
+    String json = "{\"blog-posts\":[";
+
+    for (int i = 0; i < blogPosts.size(); i++) {
+      json += gson.toJson(blogPosts.get(i));
+
+      // Adds a comma only if there is another post after the current one
+      if (i != blogPosts.size() - 1) {
+        json += ",";
+      }
+    }
+
+    json += "]}";
+    return json;
+  }
```

Comment:
Consider using high level types to reduce chance of introducing errors. Seems like this might produce the same results with less code:

```
HashMap<String, List<BlogPost>> map = // Create map with key is "blog-posts" and value is `blogPosts`
return gson.toJson(map);
```

Reply:
This is what I changed it to and it seems to be working like normal